### DISCUSSION TITLE
Is topLevelAwait supported?

### DISCUSSION THREAD

**Discussion by juanevp** [vercel/next.js#48595] [Help]
### Summary

I'm using a library which uses top level awaits in its modules. I end up importing one of those modules from middleware.ts.
If I build the app with the default config I get the error:
```Module parse failed: The top-level-await experiment is not enabled (set experiments.topLevelAwait: true to enabled it)```

When I enable this features in next.config.js and I run dev I get this error:
```The Middleware "pages/src/middleware" must export a `middleware` or a `default` function```

But the middleware module does define a:
```javascript 
export default async function middleware(req: NextRequest) {}
```

Diving into the webpack loaders I found a problem in https://github.com/vercel/next.js/blob/f078da3f314a0873dbfedbb2bb24e3412ccb7581/packages/next/src/build/webpack/loaders/next-middleware-loader.ts#L47-L48

As stated in https://webpack.js.org/configuration/experiments/, when a module uses top level await it is imported as a ```Promise```, so in the code above ```mod``` may be a ```Promise``` and that's why it won't detect the proper exports.
The question is: is this a bug or are top level awaits not supported and this is intended behavior?
Are there any known workarounds (apart from not using libraries which use the feature)?

**Comment by khuezy** [comment]
This is a bug on Nextjs' end. The fix is simple: add await around the module. https://github.com/vercel/next.js/pull/49820

**Comment by jaanaltosaar** [comment]
Also dealing with this bug @onefact! @khuezy do you know how to add await around the module please?

**Reply by khuezy** [reply]
You can patch nextjs in the 2 files in the PR above.

**Reply by khuezy** [reply]
Nextjs updated the middleware awaiting but the sandbox still doesn't properly await the `.default`. You just need to patch 1 file now: https://github.com/vercel/next.js/pull/49820

**Comment by khuezy** [comment]
They made a change to await the import of middleware code so `next dev` and build works (so it appears their intention is to support top level await in middleware), but the error now occurs at runtime, the PR above has been updated.
